nexus/mgs-updates/src/rot_bootloader_updater.rs

```diff
@@ -6,35 +6,375 @@
+                            // The naming here is a bit confusing because "stage0"
+                            // sometimes refers to the component (RoT bootloader)
+                            // and sometimes refers to the active slot for that
+                            // component. Here, we're accessing the inactive slot
+                            // for it. The component is still "stage0".
+                            &SpComponent::STAGE0.to_string(),
+                            1,
+                        )
+                        .await
+                })
+                .await;
+            let found_stage0_next_version =
+                match found_stage0_next_caboose_result {
+                    Ok(version) => {
+                        FoundVersion::Version(version.into_inner().version)
+                    }
+                    Err(error) => {
+                        if error_means_caboose_is_invalid(&error) {
+                            FoundVersion::MissingVersion
+                        } else {
+                            return Err(PrecheckError::from(error));
+                        }
+                    }
+                };
+            match (&expected_stage0_next_version, &found_stage0_next_version) {
+                // expected garbage, found garbage
+                (
+                    ExpectedVersion::NoValidVersion,
+                    FoundVersion::MissingVersion,
+                ) => (),
+                // expected a specific version and found it
+                (
+                    ExpectedVersion::Version(artifact_version),
+                    FoundVersion::Version(found_stage0_next_version),
+                ) if artifact_version.to_string()
+                    == *found_stage0_next_version =>
+                {
+                    ()
+                }
+                // anything else is a mismatch
+                (ExpectedVersion::NoValidVersion, FoundVersion::Version(_))
+                | (ExpectedVersion::Version(_), FoundVersion::MissingVersion)
+                | (ExpectedVersion::Version(_), FoundVersion::Version(_)) => {
+                    return Err(PrecheckError::WrongInactiveVersion {
+                        expected: expected_stage0_next_version.clone(),
+                        found: found_stage0_next_version,
+                    });
+                }
+            };
+
+            // The status is only considered ready for update if the stage0_next
+            // version found in the caboose is valid, and it matches what we
+            // found in the stage_0 caboose. Otherwise, we are waiting for an
+            // ongoing update.
+            match found_stage0_next_version {
+                FoundVersion::Version(v) => {
+                    if v == found_stage0_version {
+                        Ok(PrecheckStatus::ReadyForUpdate)
+                    } else {
+                        Ok(PrecheckStatus::WaitingForOngoingUpdate)
```

Comment:
I think I feel more strongly that this should be a `PrecheckError` (following up on my comment above). This feels basically the same as a version mismatch that we expect to see because another update is in progress.

Reply:
(See the other thread on this)